<a href="https://colab.research.google.com/github/ankithhhh/data-science-exercise/blob/main/Day07_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import seaborn as sns
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [ ]:
df = pd.read_csv('https://talentcocomedia.s3.amazonaws.com/ml-assets/heart-disease.csv')

In [ ]:
y = df['target']

In [ ]:
x = df.drop('target',axis=1)

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3)

##Logistic Regression - GridSearchCV

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

lr_model = LogisticRegression(penalty='l2')

val = list(np.arange(1, 10, 1))

parameters = {'solver':('newton-cg', 'liblinear', 'sag', 'saga'), 'C':val}

clf_lr = GridSearchCV(lr_model, parameters)
clf_lr.fit(x_train, y_train)

In [ ]:
print(clf_lr.best_params_)

{'C': 1, 'solver': 'liblinear'}


In [ ]:
test_pred = clf_lr.predict(x_test)

In [ ]:
acc_lr_gs = accuracy_score(test_pred,y_test)
acc_lr_gs

0.8461538461538461

Previously we had got an accuracy of 79.12% which is now improved to 84.6%

##KNeighborsClassifier - GridSearchCV

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

neigh = KNeighborsClassifier()

val = list(np.arange(1,30, 1))

parameters = {'n_neighbors':val}

clf_knn = GridSearchCV(neigh, parameters)
clf_knn.fit(x_train, y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
                                         13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
                                         23, 24, 25, 26, 27, 28, 29]})

In [ ]:
print(clf_knn.best_params_)

{'n_neighbors': 11}


In [ ]:
test_pred_knn = clf_knn.predict(x_test)

In [ ]:
accuracy_knn_gs = accuracy_score(test_pred_knn, y_test)
accuracy_knn_gs

0.6263736263736264

previously we had got an accuracy score of 60.4% which is now improved to 62.6% using GridSearchCV

##DecisionTreeClassifier - GridSearchCV

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [ ]:
clf = DecisionTreeClassifier(random_state=1024)

param_grid = {
              'max_features': ['auto', 'sqrt', 'log2'],
              'ccp_alpha': [0.1, .01, .001],
              'max_depth' : np.arange(5,10,1),
              'criterion' :['gini', 'entropy']
             }

clf_dtc = GridSearchCV(clf, param_grid=param_grid, cv=5, verbose=True)

clf_dtc.fit(x_train, y_train)

In [ ]:
print(clf_dtc.best_params_)

{'ccp_alpha': 0.01, 'criterion': 'gini', 'max_depth': 6, 'max_features': 'auto'}


In [ ]:
test_pred_dtc = clf_dtc.predict(x_test)

In [ ]:
accuracy_dtc_gs = metrics.accuracy_score(y_test, test_pred_dtc)
accuracy_dtc_gs

0.8021978021978022

previously we had got an accuracy score of 78% which is improved to 80.2% using GridSearchCV

##CONFUSION MATRIX
since we got a accuracy of 84% using Linear-Regression, we use its predicted accuracy to find the confusion matrix


##Logistic Regression

In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(test_pred,y_test)

array([[35,  3],
       [11, 42]])

In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test,test_pred))

              precision    recall  f1-score   support

           0       0.92      0.76      0.83        46
           1       0.79      0.93      0.86        45

    accuracy                           0.85        91
   macro avg       0.86      0.85      0.85        91
weighted avg       0.86      0.85      0.85        91



##KNeighborsClassifier

In [ ]:
confusion_matrix(test_pred_knn,y_test)

array([[24, 12],
       [22, 33]])

In [ ]:
print(classification_report(y_test,test_pred_knn))

              precision    recall  f1-score   support

           0       0.67      0.52      0.59        46
           1       0.60      0.73      0.66        45

    accuracy                           0.63        91
   macro avg       0.63      0.63      0.62        91
weighted avg       0.63      0.63      0.62        91



##DecisionTreeClassifier

In [ ]:
confusion_matrix(test_pred_dtc,y_test)

array([[30,  2],
       [16, 43]])

In [ ]:
print(classification_report(y_test,test_pred_dtc))

              precision    recall  f1-score   support

           0       0.94      0.65      0.77        46
           1       0.73      0.96      0.83        45

    accuracy                           0.80        91
   macro avg       0.83      0.80      0.80        91
weighted avg       0.83      0.80      0.80        91



From the above comparisions we come to a conclusion that DecisionTreeClassifier provides a better Recall along with accuracy when compared to other Modals